From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Thu, 14 Aug 2025 19:52:42 +0200
Subject: [PATCH 093/225] WIP: media: i2c: add S5K3P3 sensor driver
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Based on [1], with most parts stripped out.

TODO:
 - check if the link frequency is correct
 - decide whether to rewrite or to add the sensor to the s5k2xx driver

[1]: https://github.com/msm8953-mainline/linux/blob/6.10/main/drivers/media/i2c/s5k2xx.c

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/media/i2c/Kconfig  |    8 +
 drivers/media/i2c/Makefile |    1 +
 drivers/media/i2c/s5k3p3.c | 1085 ++++++++++++++++++++++++++++++++++++
 3 files changed, 1094 insertions(+)
 create mode 100644 drivers/media/i2c/s5k3p3.c

diff --git a/drivers/media/i2c/Kconfig b/drivers/media/i2c/Kconfig
index dd86ac983..4837f0dd2 100644
--- a/drivers/media/i2c/Kconfig
+++ b/drivers/media/i2c/Kconfig
@@ -800,6 +800,14 @@ config VIDEO_S5K3M5
 	  To compile this driver as a module, choose M here: the
 	  module will be called s5k3m5.
 
+config VIDEO_S5K3P3
+	tristate "Samsung S5K3P3 sensor support"
+	depends on I2C
+	select REGMAP_I2C
+	help
+	  This is a V4L2 sensor driver for the Samsung S5K3P3 raw
+	  camera sensor.
+
 config VIDEO_S5K5BAF
 	tristate "Samsung S5K5BAF sensor support"
 	help
diff --git a/drivers/media/i2c/Makefile b/drivers/media/i2c/Makefile
index 178822235..d3448361b 100644
--- a/drivers/media/i2c/Makefile
+++ b/drivers/media/i2c/Makefile
@@ -130,6 +130,7 @@ obj-$(CONFIG_VIDEO_RDACM21) += rdacm21.o
 obj-$(CONFIG_VIDEO_RJ54N1) += rj54n1cb0c.o
 obj-$(CONFIG_VIDEO_S5C73M3) += s5c73m3/
 obj-$(CONFIG_VIDEO_S5K3M5) += s5k3m5.o
+obj-$(CONFIG_VIDEO_S5K3P3) += s5k3p3.o
 obj-$(CONFIG_VIDEO_S5K5BAF) += s5k5baf.o
 obj-$(CONFIG_VIDEO_S5K6A3) += s5k6a3.o
 obj-$(CONFIG_VIDEO_S5KJN1) += s5kjn1.o
diff --git a/drivers/media/i2c/s5k3p3.c b/drivers/media/i2c/s5k3p3.c
new file mode 100644
index 000000000..2756c2b9b
--- /dev/null
+++ b/drivers/media/i2c/s5k3p3.c
@@ -0,0 +1,1085 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2019 Intel Corporation.
+
+#include <linux/clk.h>
+#include <linux/delay.h>
+#include <linux/gpio/consumer.h>
+#include <linux/i2c.h>
+#include <linux/module.h>
+#include <linux/of_device.h>
+#include <linux/pm_runtime.h>
+#include <linux/regmap.h>
+#include <linux/regulator/consumer.h>
+
+#include <media/v4l2-ctrls.h>
+#include <media/v4l2-device.h>
+#include <media/v4l2-fwnode.h>
+
+#define S5K3P3_MCLK_FREQ		24000000
+#define S5K3P3_DATA_LANES		4
+#define S5K3P3_MAX_COLOR_DEPTH		10
+#define S5K3P3_MAX_COLOR_VAL		(BIT(S5K3P3_MAX_COLOR_DEPTH) - 1)
+
+#define S5K3P3_REG_CHIP_ID		0x0000
+
+#define S5K3P3_REG_MODE_SELECT		0x0100
+#define S5K3P3_MODE_STANDBY		0x0000
+#define S5K3P3_MODE_STREAMING		0x0100
+
+/* vertical-timings from sensor */
+#define S5K3P3_REG_FLL			0x0340
+#define S5K3P3_FLL_30FPS		0x0e1a
+#define S5K3P3_FLL_30FPS_MIN		0x0e1a
+#define S5K3P3_FLL_MAX			0xffff
+
+/* horizontal-timings from sensor */
+#define S5K3P3_REG_LLP			0x0342
+
+/* Exposure controls from sensor */
+#define S5K3P3_REG_EXPOSURE		0x0202
+#define S5K3P3_EXPOSURE_MIN		7
+#define S5K3P3_EXPOSURE_MAX_MARGIN	8
+#define S5K3P3_EXPOSURE_STEP		1
+
+/* Analog gain controls from sensor */
+#define S5K3P3_REG_ANALOG_GAIN_MIN	0x0084
+#define S5K3P3_REG_ANALOG_GAIN_MAX	0x0086
+#define S5K3P3_REG_ANALOG_GAIN		0x0204
+#define S5K3P3_ANAL_GAIN_STEP		1
+
+/* Digital gain controls from sensor */
+#define S5K3P3_REG_DIG_GAIN		0x020e
+#define S5K3P3_DGTL_GAIN_MIN		0x100
+#define S5K3P3_DGTL_GAIN_MAX		0x1000
+#define S5K3P3_DGTL_GAIN_STEP		1
+#define S5K3P3_DGTL_GAIN_DEFAULT	0x100
+
+/* Test pattern generator */
+#define S5K3P3_REG_TEST_PATTERN		0x0600
+#define S5K3P3_REG_TEST_PATTERN_RED	0x0602
+#define S5K3P3_REG_TEST_PATTERN_GREENR	0x0604
+#define S5K3P3_REG_TEST_PATTERN_BLUE	0x0606
+#define S5K3P3_REG_TEST_PATTERN_GREENB	0x0608
+
+static const char * const s5k3p3_supply_names[] = { "avdd", "dvdd", "vio", "aux" };
+
+#define S5K3P3_NUM_SUPPLIES		ARRAY_SIZE(s5k3p3_supply_names)
+
+#define S5K3P3_LINK_FREQ_580MHZ 580500000
+#define S5K3P3_LINK_FREQ_720MHZ 720000000
+#define S5K3P3_LINK_FREQ_678MHZ 678000000
+#define S5K3P3_LINK_FREQ_1050MHZ 1050000000
+
+enum {
+	S5K3P3_LINK_FREQ_580MHZ_INDEX,
+	S5K3P3_LINK_FREQ_678MHZ_INDEX,
+	S5K3P3_LINK_FREQ_720MHZ_INDEX,
+	S5K3P3_LINK_FREQ_1050MHZ_INDEX,
+};
+
+static const s64 link_freq_menu_items[] = {
+	S5K3P3_LINK_FREQ_580MHZ,
+	S5K3P3_LINK_FREQ_678MHZ,
+	S5K3P3_LINK_FREQ_720MHZ,
+	S5K3P3_LINK_FREQ_1050MHZ,
+};
+
+struct s5k3p3_mode {
+	/* Frame width in pixels */
+	u32 width;
+
+	/* Frame height in pixels */
+	u32 height;
+
+	/* Horizontal timining size */
+	u32 llp;
+
+	/* Default vertical timining size */
+	u32 fll_def;
+
+	/* Min vertical timining size */
+	u32 fll_min;
+
+	/* Refresh rate */
+	u32 fps;
+
+	u32 link_freq_index;
+
+	/* Sensor register settings for this resolution */
+	const struct reg_sequence *regs;
+	u32 num_regs;
+};
+
+struct s5k3p3_data {
+	const char *model;
+	u32 chip_id;
+	const struct s5k3p3_mode *modes;
+	size_t num_modes;
+	const struct reg_sequence *init_regs;
+	u32 num_init_regs;
+};
+
+#define to_s5k3p3(_sd) container_of(_sd, struct s5k3p3, sd)
+
+static const char * const s5k3p3_test_pattern_menu[] = {
+	"Disabled",
+	"Solid Colour",
+	"100% Colour Bars",
+	"Fade To Grey Colour Bars",
+	"PN9",
+};
+
+static const struct reg_sequence s5k3p3_init[] = {
+	//{ 0x6028, 0x4000 },
+	//{ 0x6010, 0x0001 },
+	{ 0x6028, 0x4000 },
+	{ 0x6214, 0x7971 },
+	{ 0x6218, 0x0100 },
+	{ 0x602a, 0xf408 },
+	{ 0x6f12, 0x0048 },
+	{ 0x602a, 0xf40c },
+	{ 0x6f12, 0x0000 },
+	{ 0x602a, 0xf4aa },
+	{ 0x6f12, 0x0060 },
+	{ 0x602a, 0xf442 },
+	{ 0x6f12, 0x0800 },
+	{ 0x602a, 0xf43e },
+	{ 0x6f12, 0x0400 },
+	{ 0x6f12, 0x0000 },
+	{ 0x602a, 0xf4a4 },
+	{ 0x6f12, 0x0010 },
+	{ 0x602a, 0xf4ac },
+	{ 0x6f12, 0x0056 },
+	{ 0x602a, 0xf480 },
+	{ 0x6f12, 0x0008 },
+	{ 0x602a, 0xf492 },
+	{ 0x6f12, 0x0016 },
+	{ 0x602a, 0x3e58 },
+	{ 0x6f12, 0x0056 },
+	{ 0x602a, 0x39ee },
+	{ 0x6f12, 0x0206 },
+	{ 0x602a, 0x39e8 },
+	{ 0x6f12, 0x0205 },
+	{ 0x602a, 0x3a36 },
+	{ 0x6f12, 0xb3f0 },
+	{ 0x602a, 0x32b2 },
+	{ 0x6f12, 0x0132 },
+	{ 0x602a, 0x3a38 },
+	{ 0x6f12, 0x006c },
+	{ 0x602a, 0x3552 },
+	{ 0x6f12, 0x00d0 },
+	{ 0x602a, 0x3194 },
+	{ 0x6f12, 0x1001 },
+	{ 0x6028, 0x2000 },
+	{ 0x602a, 0x13ec },
+	{ 0x6f12, 0x8011 },
+	{ 0x6f12, 0x8011 },
+	{ 0x6028, 0x4000 },
+	{ 0x602a, 0x39ba },
+	{ 0x6f12, 0x0001 },
+	{ 0x602a, 0x3004 },
+	{ 0x6f12, 0x0008 },
+	{ 0x602a, 0x39aa },
+	{ 0x6f12, 0x0000 },
+	{ 0x6028, 0x2000 },
+	{ 0x602a, 0x026c },
+	{ 0x6f12, 0x41f0 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6028, 0x4000 },
+	{ 0x602a, 0x37d4 },
+	{ 0x6f12, 0x002d },
+	{ 0x602a, 0x37da },
+	{ 0x6f12, 0x005d },
+	{ 0x602a, 0x37e0 },
+	{ 0x6f12, 0x008d },
+	{ 0x602a, 0x37e6 },
+	{ 0x6f12, 0x00bd },
+	{ 0x602a, 0x37ec },
+	{ 0x6f12, 0x00ed },
+	{ 0x602a, 0x37f2 },
+	{ 0x6f12, 0x011d },
+	{ 0x602a, 0x37f8 },
+	{ 0x6f12, 0x014d },
+	{ 0x602a, 0x37fe },
+	{ 0x6f12, 0x017d },
+	{ 0x602a, 0x3804 },
+	{ 0x6f12, 0x01ad },
+	{ 0x602a, 0x380a },
+	{ 0x6f12, 0x01dd },
+	{ 0x602a, 0x3810 },
+	{ 0x6f12, 0x020d },
+	{ 0x602a, 0x32a6 },
+	{ 0x6f12, 0x0006 },
+	{ 0x602a, 0x32be },
+	{ 0x6f12, 0x0006 },
+	{ 0x602a, 0x3210 },
+	{ 0x6f12, 0x0006 },
+	{ 0x6028, 0x2000 },
+	{ 0x602a, 0x2ef8 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x0448 },
+	{ 0x6f12, 0x0349 },
+	{ 0x6f12, 0x0160 },
+	{ 0x6f12, 0xc26a },
+	{ 0x6f12, 0x511a },
+	{ 0x6f12, 0x8180 },
+	{ 0x6f12, 0x00f0 },
+	{ 0x6f12, 0x21b8 },
+	{ 0x6f12, 0x2000 },
+	{ 0x6f12, 0x2f78 },
+	{ 0x6f12, 0x2000 },
+	{ 0x6f12, 0x18e0 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x10b5 },
+	{ 0x6f12, 0x0e4c },
+	{ 0x6f12, 0xb4f8 },
+	{ 0x6f12, 0x5820 },
+	{ 0x6f12, 0x2388 },
+	{ 0x6f12, 0xa2eb },
+	{ 0x6f12, 0x5302 },
+	{ 0x6f12, 0xa4f8 },
+	{ 0x6f12, 0x5820 },
+	{ 0x6f12, 0x6288 },
+	{ 0x6f12, 0x5208 },
+	{ 0x6f12, 0x6280 },
+	{ 0x6f12, 0x00f0 },
+	{ 0x6f12, 0x14f8 },
+	{ 0x6f12, 0xb4f8 },
+	{ 0x6f12, 0x5800 },
+	{ 0x6f12, 0x2188 },
+	{ 0x6f12, 0x00eb },
+	{ 0x6f12, 0x5100 },
+	{ 0x6f12, 0xa4f8 },
+	{ 0x6f12, 0x5800 },
+	{ 0x6f12, 0x6088 },
+	{ 0x6f12, 0x4000 },
+	{ 0x6f12, 0x6080 },
+	{ 0x6f12, 0x10bd },
+	{ 0x6f12, 0xaff2 },
+	{ 0x6f12, 0x3300 },
+	{ 0x6f12, 0x0249 },
+	{ 0x6f12, 0x0863 },
+	{ 0x6f12, 0x7047 },
+	{ 0x6f12, 0x2000 },
+	{ 0x6f12, 0x1998 },
+	{ 0x6f12, 0x2000 },
+	{ 0x6f12, 0x0460 },
+	{ 0x6f12, 0x43f2 },
+	{ 0x6f12, 0x290c },
+	{ 0x6f12, 0xc0f2 },
+	{ 0x6f12, 0x000c },
+	{ 0x6f12, 0x6047 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x3103 },
+	{ 0x6f12, 0x0022 },
+	{ 0x6f12, 0x0000 },
+	{ 0x6f12, 0x0001 },
+};
+
+static const struct reg_sequence s5k3p3_mode_2320x1748_regs[] = {
+	{ 0x6028, 0x4000 },
+	{ 0x0344, 0x0000 },
+	{ 0x0346, 0x0000 },
+	{ 0x0348, 0x090f },
+	{ 0x034a, 0x06d3 },
+	{ 0x034c, 0x0910 },
+	{ 0x034e, 0x06d4 },
+	{ 0x3002, 0x0001 },
+	{ 0x0136, 0x1800 },
+	{ 0x0304, 0x0006 },
+	{ 0x0306, 0x008c },
+	{ 0x0302, 0x0001 },
+	{ 0x0300, 0x0008 },
+	{ 0x030c, 0x0004 },
+	{ 0x030e, 0x0078 },
+	{ 0x030a, 0x0001 },
+	{ 0x0308, 0x0008 },
+	{ 0x3008, 0x0001 },
+	{ 0x3a0c, 0x0078 },
+	{ 0x0800, 0x0000 },
+	{ 0x0200, 0x0200 },
+	{ 0x0202, 0x0100 },
+	{ 0x021c, 0x0200 },
+	{ 0x021e, 0x0100 },
+	{ 0x0342, 0x141c },
+	{ 0x0340, 0x0708 },
+	{ 0x0114, 0x0300 },
+	{ 0x3072, 0x03c0 },
+};
+
+static const struct s5k3p3_mode s5k3p3_modes[] = {
+	{
+		.width = 2320,
+		.height = 1748,
+		.fps = 30,
+		.fll_def = 1800,
+		.fll_min = 1800,
+		.llp = 5148,
+		.link_freq_index = S5K3P3_LINK_FREQ_580MHZ_INDEX,
+		.regs = s5k3p3_mode_2320x1748_regs,
+		.num_regs = ARRAY_SIZE(s5k3p3_mode_2320x1748_regs),
+	},
+};
+
+static struct s5k3p3_data s5k3p3_data = {
+	.model = "s5k3p3",
+	.chip_id = 0x3103,
+	.modes = s5k3p3_modes,
+	.num_modes = ARRAY_SIZE(s5k3p3_modes),
+	.init_regs = s5k3p3_init,
+	.num_init_regs = ARRAY_SIZE(s5k3p3_init),
+};
+
+struct s5k3p3 {
+	struct v4l2_subdev sd;
+	struct regmap *rmap;
+	struct media_pad pad;
+	struct v4l2_ctrl_handler ctrl_handler;
+
+	/* V4L2 Controls */
+	struct v4l2_ctrl *link_freq;
+	struct v4l2_ctrl *pixel_rate;
+	struct v4l2_ctrl *vblank;
+	struct v4l2_ctrl *hblank;
+	struct v4l2_ctrl *exposure;
+
+	struct regulator_bulk_data supplies[S5K3P3_NUM_SUPPLIES];
+	struct clk	 *mclk;
+	struct gpio_desc *reset_gpio;
+	int enabled;
+
+	unsigned int min_again;
+	unsigned int max_again;
+
+	/* Current mode */
+	const struct s5k3p3_mode *cur_mode;
+	const struct s5k3p3_data *data;
+
+	/* To serialize asynchronus callbacks */
+	struct mutex lock;
+
+	/* Streaming on/off */
+	bool streaming;
+};
+
+static int s5k3p3_set_ctrl(struct v4l2_ctrl *ctrl)
+{
+	struct s5k3p3 *s5k3p3 = container_of(ctrl->handler,
+					     struct s5k3p3, ctrl_handler);
+	struct i2c_client *client = v4l2_get_subdevdata(&s5k3p3->sd);
+	s64 exposure_max;
+	int ret = 0;
+
+	/* Propagate change of current control to all related controls */
+	if (ctrl->id == V4L2_CID_VBLANK) {
+		/* Update max exposure while meeting expected vblanking */
+		exposure_max = s5k3p3->cur_mode->height + ctrl->val -
+			       S5K3P3_EXPOSURE_MAX_MARGIN;
+		__v4l2_ctrl_modify_range(s5k3p3->exposure,
+					 s5k3p3->exposure->minimum,
+					 exposure_max, s5k3p3->exposure->step,
+					 exposure_max);
+	}
+
+	/* V4L2 controls values will be applied only when power is already up */
+	if (!pm_runtime_get_if_in_use(&client->dev))
+		return 0;
+
+	switch (ctrl->id) {
+	case V4L2_CID_EXPOSURE:
+		ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_EXPOSURE,
+				ctrl->val);
+		break;
+	case V4L2_CID_ANALOGUE_GAIN:
+		ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_ANALOG_GAIN,
+				ctrl->val);
+		break;
+	case V4L2_CID_DIGITAL_GAIN:
+		ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_DIG_GAIN,
+				ctrl->val);
+		break;
+	case V4L2_CID_VBLANK:
+		ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_FLL,
+				s5k3p3->cur_mode->height + ctrl->val);
+		break;
+	case V4L2_CID_TEST_PATTERN:
+		ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_TEST_PATTERN,
+				ctrl->val);
+		break;
+	case V4L2_CID_TEST_PATTERN_RED:
+		ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_TEST_PATTERN_RED,
+				ctrl->val);
+		break;
+	case V4L2_CID_TEST_PATTERN_GREENR:
+		ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_TEST_PATTERN_GREENR,
+				ctrl->val);
+		break;
+	case V4L2_CID_TEST_PATTERN_BLUE:
+		ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_TEST_PATTERN_BLUE,
+				ctrl->val);
+		break;
+	case V4L2_CID_TEST_PATTERN_GREENB:
+		ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_TEST_PATTERN_GREENB,
+				ctrl->val);
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+
+	pm_runtime_put(&client->dev);
+
+	return ret;
+}
+
+static const struct v4l2_ctrl_ops s5k3p3_ctrl_ops = {
+	.s_ctrl = s5k3p3_set_ctrl,
+};
+
+static s64 calc_pixel_rate(const struct s5k3p3_mode *mode)
+{
+	u64 pixel_rate = link_freq_menu_items[mode->link_freq_index] * 2 * S5K3P3_DATA_LANES;
+
+	do_div(pixel_rate, S5K3P3_MAX_COLOR_DEPTH);
+
+	return pixel_rate;
+}
+
+static int s5k3p3_init_controls(struct s5k3p3 *s5k3p3)
+{
+	struct i2c_client *client = v4l2_get_subdevdata(&s5k3p3->sd);
+	struct v4l2_ctrl_handler *ctrl_hdlr;
+	struct v4l2_fwnode_device_properties props;
+	s64 exposure_max, h_blank;
+	s64 rate_max = INT_MIN, rate_min = INT_MAX;
+	int ret, i;
+
+	ctrl_hdlr = &s5k3p3->ctrl_handler;
+	ret = v4l2_ctrl_handler_init(ctrl_hdlr, 8);
+	if (ret)
+		return ret;
+
+	ctrl_hdlr->lock = &s5k3p3->lock;
+
+	for (i = 0; i < s5k3p3->data->num_modes; i++) {
+		rate_max = max(rate_max, calc_pixel_rate(&s5k3p3->data->modes[i]));
+		rate_min = min(rate_min, calc_pixel_rate(&s5k3p3->data->modes[i]));
+	}
+
+	s5k3p3->link_freq = v4l2_ctrl_new_int_menu(ctrl_hdlr, &s5k3p3_ctrl_ops,
+						  V4L2_CID_LINK_FREQ,
+					ARRAY_SIZE(link_freq_menu_items) - 1,
+					0, link_freq_menu_items);
+	if (s5k3p3->link_freq)
+		s5k3p3->link_freq->flags |= V4L2_CTRL_FLAG_READ_ONLY;
+
+	s5k3p3->pixel_rate = v4l2_ctrl_new_std (ctrl_hdlr, &s5k3p3_ctrl_ops,
+			V4L2_CID_PIXEL_RATE, rate_min, rate_max,
+			1, calc_pixel_rate(s5k3p3->cur_mode));
+
+	s5k3p3->vblank = v4l2_ctrl_new_std(ctrl_hdlr, &s5k3p3_ctrl_ops,
+					  V4L2_CID_VBLANK,
+					  s5k3p3->cur_mode->fll_min -
+					  s5k3p3->cur_mode->height,
+					  S5K3P3_FLL_MAX -
+					  s5k3p3->cur_mode->height, 1,
+					  s5k3p3->cur_mode->fll_def -
+					  s5k3p3->cur_mode->height);
+
+	h_blank = s5k3p3->cur_mode->llp - s5k3p3->cur_mode->width;
+
+	s5k3p3->hblank = v4l2_ctrl_new_std(ctrl_hdlr, &s5k3p3_ctrl_ops,
+					  V4L2_CID_HBLANK, h_blank, h_blank, 1,
+					  h_blank);
+	if (s5k3p3->hblank)
+		s5k3p3->hblank->flags |= V4L2_CTRL_FLAG_READ_ONLY;
+
+	v4l2_ctrl_new_std(ctrl_hdlr, &s5k3p3_ctrl_ops, V4L2_CID_ANALOGUE_GAIN,
+			  s5k3p3->min_again, s5k3p3->max_again,
+			  S5K3P3_ANAL_GAIN_STEP, s5k3p3->min_again);
+	v4l2_ctrl_new_std(ctrl_hdlr, &s5k3p3_ctrl_ops, V4L2_CID_DIGITAL_GAIN,
+			  S5K3P3_DGTL_GAIN_MIN, S5K3P3_DGTL_GAIN_MAX,
+			  S5K3P3_DGTL_GAIN_STEP, S5K3P3_DGTL_GAIN_DEFAULT);
+	exposure_max = s5k3p3->cur_mode->fll_def - S5K3P3_EXPOSURE_MAX_MARGIN;
+	s5k3p3->exposure = v4l2_ctrl_new_std(ctrl_hdlr, &s5k3p3_ctrl_ops,
+					    V4L2_CID_EXPOSURE,
+					    S5K3P3_EXPOSURE_MIN, exposure_max,
+					    S5K3P3_EXPOSURE_STEP,
+					    exposure_max);
+
+	v4l2_ctrl_new_std_menu_items(ctrl_hdlr, &s5k3p3_ctrl_ops,
+				     V4L2_CID_TEST_PATTERN,
+				     ARRAY_SIZE(s5k3p3_test_pattern_menu) - 1,
+				     0, 0, s5k3p3_test_pattern_menu);
+
+	for (i = 0; i < 4; i++) {
+		v4l2_ctrl_new_std(ctrl_hdlr, &s5k3p3_ctrl_ops,
+				  V4L2_CID_TEST_PATTERN_RED + i,
+				  0, S5K3P3_MAX_COLOR_VAL,
+				  1, S5K3P3_MAX_COLOR_VAL);
+	}
+
+	ret = v4l2_fwnode_device_parse(&client->dev, &props);
+	if (ret)
+		return ret;
+
+	ret = v4l2_ctrl_new_fwnode_properties(ctrl_hdlr, &s5k3p3_ctrl_ops,
+					      &props);
+	if (ret)
+		return ret;
+
+	if (ctrl_hdlr->error)
+		return ctrl_hdlr->error;
+
+	s5k3p3->sd.ctrl_handler = ctrl_hdlr;
+
+	return 0;
+}
+
+static void s5k3p3_assign_pad_format(const struct s5k3p3_mode *mode,
+				    struct v4l2_mbus_framefmt *fmt)
+{
+	fmt->width = mode->width;
+	fmt->height = mode->height;
+	fmt->code = MEDIA_BUS_FMT_SGRBG10_1X10;
+	fmt->field = V4L2_FIELD_NONE;
+}
+
+static int s5k3p3_start_streaming(struct s5k3p3 *s5k3p3)
+{
+	struct i2c_client *client = v4l2_get_subdevdata(&s5k3p3->sd);
+	int ret;
+
+	ret = regmap_multi_reg_write(s5k3p3->rmap, s5k3p3->data->init_regs,
+						   s5k3p3->data->num_init_regs);
+	if (ret) {
+		dev_err(&client->dev, "failed to set plls: %d", ret);
+		return ret;
+	}
+
+	ret = regmap_multi_reg_write(s5k3p3->rmap, s5k3p3->cur_mode->regs,
+						   s5k3p3->cur_mode->num_regs);
+	if (ret) {
+		dev_err(&client->dev, "failed to set mode: %d", ret);
+		return ret;
+	}
+
+	ret = __v4l2_ctrl_handler_setup(s5k3p3->sd.ctrl_handler);
+	if (ret)
+		return ret;
+
+	ret = regmap_write(s5k3p3->rmap, S5K3P3_REG_MODE_SELECT,
+			      S5K3P3_MODE_STREAMING);
+
+	if (ret) {
+		dev_err(&client->dev, "failed to set stream");
+		return ret;
+	}
+
+	return 0;
+}
+
+static void s5k3p3_stop_streaming(struct s5k3p3 *s5k3p3)
+{
+	struct i2c_client *client = v4l2_get_subdevdata(&s5k3p3->sd);
+
+	if (regmap_write(s5k3p3->rmap, S5K3P3_REG_MODE_SELECT,
+			    S5K3P3_MODE_STANDBY))
+		dev_err(&client->dev, "failed to set stream");
+}
+
+static int s5k3p3_set_stream(struct v4l2_subdev *sd, int enable)
+{
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+	struct i2c_client *client = v4l2_get_subdevdata(sd);
+	int ret = 0;
+
+	if (s5k3p3->streaming == enable)
+		return 0;
+
+	mutex_lock(&s5k3p3->lock);
+	if (enable) {
+		ret = pm_runtime_get_sync(&client->dev);
+		if (ret < 0) {
+			pm_runtime_put_noidle(&client->dev);
+			mutex_unlock(&s5k3p3->lock);
+			return ret;
+		}
+
+		ret = s5k3p3_start_streaming(s5k3p3);
+		if (ret) {
+			enable = 0;
+			s5k3p3_stop_streaming(s5k3p3);
+			pm_runtime_put(&client->dev);
+		}
+	} else {
+		s5k3p3_stop_streaming(s5k3p3);
+		pm_runtime_put(&client->dev);
+	}
+
+	s5k3p3->streaming = enable;
+	mutex_unlock(&s5k3p3->lock);
+
+	return ret;
+}
+
+static int __maybe_unused s5k3p3_suspend(struct device *dev)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+
+	mutex_lock(&s5k3p3->lock);
+	if (s5k3p3->streaming)
+		s5k3p3_stop_streaming(s5k3p3);
+
+	mutex_unlock(&s5k3p3->lock);
+
+	return 0;
+}
+
+static int __maybe_unused s5k3p3_resume(struct device *dev)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+	int ret;
+
+	mutex_lock(&s5k3p3->lock);
+	if (s5k3p3->streaming) {
+		ret = s5k3p3_start_streaming(s5k3p3);
+		if (ret)
+			goto error;
+	}
+
+	mutex_unlock(&s5k3p3->lock);
+
+	return 0;
+
+error:
+	s5k3p3_stop_streaming(s5k3p3);
+	s5k3p3->streaming = 0;
+	mutex_unlock(&s5k3p3->lock);
+	return ret;
+}
+
+static int s5k3p3_power_on(struct device *dev)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+	int ret;
+
+	if ((s5k3p3->enabled)++)
+		return 0;
+
+	ret = regulator_bulk_enable(S5K3P3_NUM_SUPPLIES,
+				    s5k3p3->supplies);
+	if (ret) {
+		dev_err(&client->dev, "%s: failed to enable regulators\n",
+			__func__);
+		return ret;
+	}
+
+	ret = clk_prepare_enable(s5k3p3->mclk);
+	if (ret) {
+		dev_err(&client->dev, "%s: failed to enable clock\n",
+			__func__);
+		goto reg_off;
+	}
+
+	usleep_range(1000, 2000);
+
+	gpiod_set_value_cansleep(s5k3p3->reset_gpio, 0);
+
+	usleep_range(10000, 11000);
+
+	return 0;
+
+reg_off:
+	regulator_bulk_disable(S5K3P3_NUM_SUPPLIES, s5k3p3->supplies);
+
+	return ret;
+}
+
+static int s5k3p3_power_off(struct device *dev)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+
+	if (--(s5k3p3->enabled) > 0)
+		return 0;
+
+	gpiod_set_value_cansleep(s5k3p3->reset_gpio, 1);
+
+	regulator_bulk_disable(S5K3P3_NUM_SUPPLIES, s5k3p3->supplies);
+	clk_disable_unprepare(s5k3p3->mclk);
+
+	return 0;
+}
+
+static int s5k3p3_set_format(struct v4l2_subdev *sd,
+			    struct v4l2_subdev_state *sd_state,
+			    struct v4l2_subdev_format *fmt)
+{
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+	const struct s5k3p3_mode *mode;
+	s32 vblank_def, h_blank;
+
+	mode = v4l2_find_nearest_size(s5k3p3->data->modes,
+				      s5k3p3->data->num_modes, width,
+				      height, fmt->format.width,
+				      fmt->format.height);
+
+	mutex_lock(&s5k3p3->lock);
+	s5k3p3_assign_pad_format(mode, &fmt->format);
+	if (fmt->which == V4L2_SUBDEV_FORMAT_TRY) {
+		*v4l2_subdev_state_get_format(sd_state, fmt->pad) = fmt->format;
+	} else {
+		s5k3p3->cur_mode = mode;
+		__v4l2_ctrl_s_ctrl(s5k3p3->link_freq, mode->link_freq_index);
+
+		__v4l2_ctrl_s_ctrl_int64(s5k3p3->pixel_rate,
+					 calc_pixel_rate(mode));
+
+		/* Update limits and set FPS to default */
+		vblank_def = mode->fll_def - mode->height;
+		__v4l2_ctrl_modify_range(s5k3p3->vblank,
+					 mode->fll_min - mode->height,
+					 S5K3P3_FLL_MAX - mode->height, 1,
+					 vblank_def);
+		__v4l2_ctrl_s_ctrl(s5k3p3->vblank, vblank_def);
+
+		h_blank = s5k3p3->cur_mode->llp - s5k3p3->cur_mode->width;
+
+		__v4l2_ctrl_modify_range(s5k3p3->hblank, h_blank, h_blank, 1,
+					 h_blank);
+	}
+
+	mutex_unlock(&s5k3p3->lock);
+
+	return 0;
+}
+
+static int s5k3p3_get_format(struct v4l2_subdev *sd,
+			    struct v4l2_subdev_state *sd_state,
+			    struct v4l2_subdev_format *fmt)
+{
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+
+	mutex_lock(&s5k3p3->lock);
+	if (fmt->which == V4L2_SUBDEV_FORMAT_TRY)
+		fmt->format = *v4l2_subdev_state_get_format(sd_state,
+							  fmt->pad);
+	else
+		s5k3p3_assign_pad_format(s5k3p3->cur_mode, &fmt->format);
+
+	mutex_unlock(&s5k3p3->lock);
+
+	return 0;
+}
+
+static int s5k3p3_enum_mbus_code(struct v4l2_subdev *sd,
+				struct v4l2_subdev_state *sd_state,
+				struct v4l2_subdev_mbus_code_enum *code)
+{
+	if (code->index > 0)
+		return -EINVAL;
+
+	code->code = MEDIA_BUS_FMT_SGRBG10_1X10;
+
+	return 0;
+}
+
+static int s5k3p3_enum_frame_size(struct v4l2_subdev *sd,
+				 struct v4l2_subdev_state *sd_state,
+				 struct v4l2_subdev_frame_size_enum *fse)
+{
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+
+	if (fse->index >= s5k3p3->data->num_modes)
+		return -EINVAL;
+
+	if (fse->code != MEDIA_BUS_FMT_SGRBG10_1X10)
+		return -EINVAL;
+
+	fse->min_width = s5k3p3->data->modes[fse->index].width;
+	fse->max_width = fse->min_width;
+	fse->min_height = s5k3p3->data->modes[fse->index].height;
+	fse->max_height = fse->min_height;
+
+	return 0;
+}
+
+static int s5k3p3_open(struct v4l2_subdev *sd, struct v4l2_subdev_fh *fh)
+{
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+
+	mutex_lock(&s5k3p3->lock);
+	s5k3p3_assign_pad_format(&s5k3p3->data->modes[0],
+				v4l2_subdev_state_get_format(fh->state, 0));
+	mutex_unlock(&s5k3p3->lock);
+
+	return 0;
+}
+
+static const struct v4l2_subdev_video_ops s5k3p3_video_ops = {
+	.s_stream = s5k3p3_set_stream,
+};
+
+static const struct v4l2_subdev_pad_ops s5k3p3_pad_ops = {
+	.set_fmt = s5k3p3_set_format,
+	.get_fmt = s5k3p3_get_format,
+	.enum_mbus_code = s5k3p3_enum_mbus_code,
+	.enum_frame_size = s5k3p3_enum_frame_size,
+};
+
+static const struct v4l2_subdev_ops s5k3p3_subdev_ops = {
+	.video = &s5k3p3_video_ops,
+	.pad = &s5k3p3_pad_ops,
+};
+
+static const struct media_entity_operations s5k3p3_subdev_entity_ops = {
+	.link_validate = v4l2_subdev_link_validate,
+};
+
+static const struct v4l2_subdev_internal_ops s5k3p3_internal_ops = {
+	.open = s5k3p3_open,
+};
+
+static int s5k3p3_identify_module(struct s5k3p3 *s5k3p3)
+{
+	struct i2c_client *client = v4l2_get_subdevdata(&s5k3p3->sd);
+	int ret;
+	u32 val;
+
+	ret = regmap_read(s5k3p3->rmap, S5K3P3_REG_CHIP_ID, &val);
+	if (ret)
+		return ret;
+
+	if (val != s5k3p3->data->chip_id) {
+		dev_err(&client->dev, "chip id mismatch: %x!=%x",
+			s5k3p3->data->chip_id, val);
+		return -ENXIO;
+	}
+
+	ret = regmap_read(s5k3p3->rmap, S5K3P3_REG_ANALOG_GAIN_MIN, &val);
+	if (ret)
+		return ret;
+
+	s5k3p3->min_again = val;
+
+	ret = regmap_read(s5k3p3->rmap, S5K3P3_REG_ANALOG_GAIN_MAX, &val);
+	if (ret)
+		return ret;
+
+	s5k3p3->max_again = val;
+
+	dev_info(&client->dev, "Analog gain: min=%u, max=%u", s5k3p3->min_again, val);
+
+	return 0;
+}
+
+static int s5k3p3_check_hwcfg(struct s5k3p3 *s5k3p3, struct device *dev)
+{
+	struct fwnode_handle *ep;
+	struct fwnode_handle *fwnode = dev_fwnode(dev);
+	struct v4l2_fwnode_endpoint bus_cfg = {
+		.bus_type = V4L2_MBUS_CSI2_DPHY
+	};
+	int ret = 0;
+
+	if (!fwnode)
+		return -ENXIO;
+
+	ep = fwnode_graph_get_next_endpoint(fwnode, NULL);
+	if (!ep)
+		return -ENXIO;
+
+	ret = v4l2_fwnode_endpoint_alloc_parse(ep, &bus_cfg);
+	fwnode_handle_put(ep);
+	if (ret)
+		return ret;
+
+	if (bus_cfg.bus.mipi_csi2.num_data_lanes != S5K3P3_DATA_LANES) {
+		dev_err(dev, "number of CSI2 data lanes %d is not supported",
+			bus_cfg.bus.mipi_csi2.num_data_lanes);
+		ret = -EINVAL;
+	}
+
+	v4l2_fwnode_endpoint_free(&bus_cfg);
+
+	return ret;
+}
+
+static void s5k3p3_remove(struct i2c_client *client)
+{
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct s5k3p3 *s5k3p3 = to_s5k3p3(sd);
+
+	v4l2_async_unregister_subdev(sd);
+	media_entity_cleanup(&sd->entity);
+	v4l2_ctrl_handler_free(sd->ctrl_handler);
+	pm_runtime_disable(&client->dev);
+	mutex_destroy(&s5k3p3->lock);
+}
+
+static int s5k3p3_of_init(struct s5k3p3 *s5k3p3, struct device *dev)
+{
+	u32 mclk_freq;
+	int i, ret;
+
+	if (!dev->of_node)
+		return 0;
+
+	s5k3p3->data = (struct s5k3p3_data*) of_device_get_match_data(dev);
+
+	for (i = 0; i < S5K3P3_NUM_SUPPLIES; i++)
+		s5k3p3->supplies[i].supply = s5k3p3_supply_names[i];
+
+	ret = devm_regulator_bulk_get(dev, S5K3P3_NUM_SUPPLIES, s5k3p3->supplies);
+	if (ret < 0)
+		return ret;
+
+	s5k3p3->mclk = devm_clk_get(dev, NULL);
+	if (IS_ERR(s5k3p3->mclk))
+		return PTR_ERR(s5k3p3->mclk);
+
+	mclk_freq = clk_get_rate(s5k3p3->mclk);
+	if (mclk_freq != S5K3P3_MCLK_FREQ) {
+		dev_err(dev, "external clock frequency %u is not supported",
+			mclk_freq);
+		return -EINVAL;
+	}
+
+	/* Request optional enable pin */
+	s5k3p3->reset_gpio = devm_gpiod_get_optional(dev, "reset", GPIOD_OUT_HIGH);
+	if (IS_ERR(s5k3p3->reset_gpio))
+		return PTR_ERR(s5k3p3->reset_gpio);
+
+	return 0;
+}
+
+static const struct regmap_config s5k3p3_regmap_config = {
+	.reg_bits = 16,
+	.val_bits = 16,
+	.cache_type = REGCACHE_NONE,
+};
+
+static int s5k3p3_probe(struct i2c_client *client)
+{
+	struct s5k3p3 *s5k3p3;
+	int ret;
+
+	s5k3p3 = devm_kzalloc(&client->dev, sizeof(*s5k3p3), GFP_KERNEL);
+	if (!s5k3p3)
+		return -ENOMEM;
+
+	ret = s5k3p3_of_init(s5k3p3, &client->dev);
+	if (ret)
+		return ret;
+
+	ret = s5k3p3_check_hwcfg(s5k3p3, &client->dev);
+	if (ret) {
+		dev_err(&client->dev, "failed to check HW configuration: %d",
+			ret);
+		return ret;
+	}
+
+	s5k3p3->rmap = devm_regmap_init_i2c(client, &s5k3p3_regmap_config);
+	if (IS_ERR_OR_NULL(s5k3p3->rmap)) {
+		dev_err(&client->dev, "failed to initialize regmap: %ld\n",
+			PTR_ERR(s5k3p3->rmap));
+		return PTR_ERR(s5k3p3->rmap) ?: -ENODATA;
+	}
+
+	v4l2_i2c_subdev_init(&s5k3p3->sd, client, &s5k3p3_subdev_ops);
+
+	s5k3p3_power_on(&client->dev);
+
+	ret = s5k3p3_identify_module(s5k3p3);
+	if (ret) {
+		dev_err(&client->dev, "failed to find sensor: %d", ret);
+		goto power_off;
+	}
+
+	mutex_init(&s5k3p3->lock);
+	s5k3p3->cur_mode = &s5k3p3->data->modes[0];
+	ret = s5k3p3_init_controls(s5k3p3);
+	if (ret) {
+		dev_err(&client->dev, "failed to init controls: %d", ret);
+		goto probe_error_v4l2_ctrl_handler_free;
+	}
+
+	snprintf(s5k3p3->sd.name, sizeof(s5k3p3->sd.name),
+			"%s %s", s5k3p3->data->model, dev_name(&client->dev));
+
+	s5k3p3->sd.internal_ops = &s5k3p3_internal_ops;
+	s5k3p3->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE;
+	s5k3p3->sd.entity.ops = &s5k3p3_subdev_entity_ops;
+	s5k3p3->sd.entity.function = MEDIA_ENT_F_CAM_SENSOR;
+	s5k3p3->pad.flags = MEDIA_PAD_FL_SOURCE;
+	ret = media_entity_pads_init(&s5k3p3->sd.entity, 1, &s5k3p3->pad);
+	if (ret) {
+		dev_err(&client->dev, "failed to init entity pads: %d", ret);
+		goto probe_error_v4l2_ctrl_handler_free;
+	}
+
+	ret = v4l2_async_register_subdev_sensor(&s5k3p3->sd);
+	if (ret < 0) {
+		dev_err(&client->dev, "failed to register V4L2 subdev: %d",
+			ret);
+		goto probe_error_media_entity_cleanup;
+	}
+
+	pm_runtime_set_active(&client->dev);
+	pm_runtime_enable(&client->dev);
+	pm_runtime_idle(&client->dev);
+
+	return 0;
+
+probe_error_media_entity_cleanup:
+	media_entity_cleanup(&s5k3p3->sd.entity);
+
+probe_error_v4l2_ctrl_handler_free:
+	v4l2_ctrl_handler_free(s5k3p3->sd.ctrl_handler);
+	mutex_destroy(&s5k3p3->lock);
+
+power_off:
+	s5k3p3_power_off(&client->dev);
+
+	return ret;
+}
+
+static const struct dev_pm_ops s5k3p3_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(s5k3p3_suspend, s5k3p3_resume)
+	SET_RUNTIME_PM_OPS(s5k3p3_power_off, s5k3p3_power_on, NULL)
+};
+
+static const struct of_device_id s5k3p3_of_match[] = {
+	{ .compatible = "samsung,s5k3p3", &s5k3p3_data },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, s5k3p3_of_match);
+
+static struct i2c_driver s5k3p3_i2c_driver = {
+	.driver = {
+		.name = "s5k3p3",
+		.pm = &s5k3p3_pm_ops,
+		.of_match_table	= of_match_ptr(s5k3p3_of_match),
+	},
+	.probe = s5k3p3_probe,
+	.remove = s5k3p3_remove,
+};
+
+module_i2c_driver(s5k3p3_i2c_driver);
+
+MODULE_DESCRIPTION("Samsung S5K3P3 sensor driver");
+MODULE_LICENSE("GPL v2");
